### ISSUE TITLE
Specifying compopts etc. in comments in tests?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#6658]
This issue is a spin-off of #6628.

In particular the proposal is that, besides specifying that a test is to use FileCheck in a comment, you can also specify compopts, execopts, noexec, and numlocales.

All of these special comments must start with // and go at the top of the file.

E.g.

// COMPOPTS: --baseline
writeln("Baseline Hello World");

or

// FileCheck
// COMPOPTS: --llvm --llvm-print-ir test --llvm-print-ir-stage none
proc test() {
  writeln("LLVM IR Hello World");
}

These comment forms of compopts/execopts/etc take the place of the sometest.compopts. It would be an error to have both for a particular test. The keywords are case-insensitive.

**Comment by mppf** [comment]
### Why do this?

It seems like an improvement to allow FileCheck-based new tests to be written with only .chpl files.

### Will it make it harder to look for compiler flags in tests?

We already have quite a bit of places to look for these because:
 * COMPOPTS vs test.compopts
 * performance testing uses different files
 * compopts can be specified on the start_test / paratest command line

As far as grepping goes, we don't currently have a script to check for a pattern just in .compopts files. For that reason I usually just do something like `grep -r some-compiler-flag test`. Such an approach would find the COMPOPTS comments just fine. If we were to add a `grepcompopts` script, it would be trivial to make it also check for COMPOPTS comments.

### Should we support all the sometest.xyz files in this form?

No. In particular I wouldn't expect executable .compopts (say) to move to comments. Furthermore, my goal here was to just make the most-used configurations available.

**Comment by coodie** [comment]
I really like that idea because it is way much more convenient to write tests and make them much more readable, if it is possible to keep everything in one file, then do it., from my point of view having COMPOPTS as a comment instead of separate file is much more convenient from point of test writer/maintainer. Nothing related to changing compopts came to my mind while thinking about adding FileCheck, but when I see this proposition I'd love to have it. I browse LLVM's and Clang tests quite often (since it's related to my GSoC project) and I must say it is definitely easier to browse and read their tests than chapel's, because everything is contained in one file: configuration (using RUN commands), file to be tested (passed to Clang/Opt/whatever), expected output (FileCheck commands). While in case of chapel I have to read ".prediff" bash script for every test that does something slightly more complicated, see .good file, .compopts, .skipif for potential test configuration, having to write/read this many files makes adding/maintaining tests much more complicated that it could potentially be.

It's not a big deal to add .compopts file per each test, or even open .compopts and .chpl file to see both of them simultaneously in your editor if you want to have both of these on your screen, but having all of that in one file is very convenient. One VERY important thing to note is that developers in general don't like writing tests, while they are very important. We want to have good quality tests and relieve some pain from potentially frustrated developer that "has to write them again", if there is any way to make it more convenient I'm 100% sure it will have big impact on quality of testing and their maintenance.

**Comment by lydia-duncan** [comment]
Note that in the past we have discussed moving our test settings to a YAML-based single file format for precisely these reasons, but haven't had the time to get around to it.  Based on that (admittedly brief) description would the FileCheck format be compatible with that goal, an eventual replacement for that strategy, or another mode entirely?

**Comment by mppf** [comment]
@lydia-duncan - I don't know, for one thing I'm not familiar with YAML and wasn't part of that discussion. However, I think it'd be reasonable to support simple settings with a `// COMPOPTS:` style whether or not YAML is involved for more complex things. And anyway, I personally don't see the reason to invoke YAML or JSON or TOML or XML or .... for specifying test options.

In any case, based upon a cursory examination of a YAML sample, it appears to use a `key: value` format which happens to be the same as what I'm proposing. So, I'd guess that it is compatible with a YAML future if we choose to go there.

**Comment by vasslitvinov** [comment]
I like Michael's idea, at the same time we need to consider our earlier YAML design.

There, one point was that all setup things are specified in one file. This proposal takes it one step further by merging that with the Chapel source file.

More importantly, YAML format allows specification of multiple configurations. For example, "when testing under gasnet, use these two sets of compiler options; the .good file for the first set is XXX, for the second YYY". Combining compiler and executable options and specifying .good files for each combination; coordinating the settings for correctness and performance runs; etc. While there, specifying common prediffs with just a Python function call.

I am OK with Michael's suggestion that for now we go for allowing simple specs as comments in .chpl file.

Actually, when/if we enact a yaml-based scheme, we may still want to keep the ability to provide specs as comments in .chpl as a simpler alternative to a yaml file.

(Looking forward, how about allowing .good output as a comment in .chpl file as well?)

**Comment by noakesmichael** [comment]
I do not believe that Chapel source files should have text in comments that impact the
behavior of the compilation and/or execution environment.

**Comment by bradcray** [comment]
@noakesmichael : Just to make sure we're on the same page: As I understand it, these comments would only affect the behavior of the tests when run in the Chapel test system, not when run through the compiler using a normal command-line call like `chpl foo.chpl`.  Does your objection still stand?  Thanks.

**Comment by mppf** [comment]
@bradcray - FYI @ronawho didn't like this idea either. I'll leave it to him to elaborate on the objection but at the moment it doesn't look like this proposal has sufficient support.

**Comment by ronawho** [comment]
I was under the impression that comment compopts and file-level .compopts would work together, which I thought we be confusing, but in the issue description it says " It would be an error to have both for a particular test", so I don't object as much.
